eth/api_tracer.go

```diff
@@ -802,19 +811,100 @@ func (api *PrivateDebugAPI) TraceTransaction(ctx context.Context, hash common.Ha
 	return traceTransaction(ctx, api.eth, hash, config)
 }
 
+// traceCall lets you trace a given eth_call. It collects the structured logs created during the execution of EVM
+// if the given transaction was added on top of the provided block and returns them as a JSON object.
+// You can provide -2 as a block number to trace on top of the pending block.
+func traceCall(ctx context.Context, eth *Ethereum, args ethapi.CallArgs, blockNrOrHash rpc.BlockNumberOrHash, config *TraceConfig) (interface{}, error) {
+	// First try to retrieve the state
+	blockNrOrHash.RequireCanonical = true
+	statedb, header, err := eth.APIBackend.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	if err != nil {
+		// Try to retrieve the specified block
+		var block *types.Block
+		if hash, ok := blockNrOrHash.Hash(); ok {
+			block = eth.blockchain.GetBlockByHash(hash)
+		} else if number, ok := blockNrOrHash.Number(); ok {
+			block = eth.blockchain.GetBlockByNumber(uint64(number))
+		}
+		if block == nil {
+			return nil, fmt.Errorf("block %v not found: %v", blockNrOrHash, err)
+		}
+		// try to recompute the state
+		reexec := defaultTraceReexec
+		if config != nil && config.Reexec != nil {
+			reexec = *config.Reexec
+		}
+		_, _, statedb, err = computeTxEnv(eth, block, len(block.Transactions())-1, reexec)
```

Comment:
IIRC, this `len(block.txs)-1` (vs. `0`) was related to a bug found at ethereum/go-ethereum w/r/t the computed env, or? Just out of curiousity, did report that or they get it fixed?

Reply:
Yes, that's indeed a bug that we handled here. From what I recall they handled this fine in the trace package refactor, but I will have to check for sure.

It's on my list of things to refer/PR upstream.

Reply:
@meowsbits this has been resolved upstream.